src/Swarm/Game/Step.hs

```diff
@@ -697,30 +716,63 @@ execConst c vs k = do
     Make -> case vs of
       [VString name] -> do
         inv <- use robotInventory
-        em <- lift . lift $ use entityMap
+        ins <- use installedDevices
+        em <- doOnGame $ use entityMap
         e <-
           lookupEntityName name em
             `isJustOr` cmdExn Make ["I've never heard of", indefiniteQ name, "."]
 
-        outRs <- lift . lift $ use recipesOut
+        outRs <- doOnGame $ use recipesOut
 
         -- Only consider recipes where the number of things we are trying to make
         -- is greater in the outputs than in the inputs.  This prevents us from doing
         -- silly things like making copper pipes when the user says "make furnace".
         let recipes = filter increase (recipesFor outRs e)
-            increase (Recipe ins outs _) = countIn outs > countIn ins
+            increase r = countIn (r ^. recipeOutputs) > countIn (r ^. recipeInputs)
             countIn xs = maybe 0 fst (find ((== e) . snd) xs)
         not (null recipes)
           `holdsOr` cmdExn Make ["There is no known recipe for making", indefinite name, "."]
 
         -- Now try each recipe and take the first one that we have the
         -- ingredients for.
-        inv' <-
-          listToMaybe (rights (map (make inv) recipes))
+        let makeRecipe r = (,r) <$> make (inv, ins) r
+        (inv', recipe) <-
+          listToMaybe (rights (map makeRecipe recipes))
             `isJustOr` cmdExn Make ["You don't have the ingredients to make", indefinite name, "."]
 
-        robotInventory .= inv'
-        return $ Out VUnit k
+        finishCookingRecipe recipe (WorldFun Right) (RobotFun $ const inv')
```

Comment:
I don't think `const inv'` is right.  What if another robot `give`s this robot something while it is waiting for its recipe to complete?  When the recipe completes the inventory will be replaced with `inv'` and we will lose the thing that the other robot gave.  Instead I think we just have to remember some things to delete and some things to add.

Reply:
Indeed, we should probably make this an official invariant: stuff cannot *leave* an active robot's inventory unless under its explicit control.